logstash-core/lib/logstash/codecs/base.rb

```diff
@@ -28,10 +29,56 @@ def decode(data)
   alias_method :<<, :decode
 
   public
+  # DEPRECATED: Prefer defining encode_sync or multi_encode
   def encode(event)
-    raise "#{self.class}#encode must be overidden"
+    @on_event.call(event, encode_sync(event))
   end # def encode
 
+  # This is a preferred way (along with multiencode) to define a codec's encode function.
+  # This default implementation should really never be invoked by an output plugin
+  # this implementation merely provided for completeness. Plugins should prefer to simply use the
+  # concurrent_multi_receive API which always invokes multi_encode.
+  def encode_sync(event)
+    event, data = multi_encode([event]).first
+    data
+  end
+
+  public
+  # Relies on the codec being synchronous (which they all are!)
+  # We need a better long term design here, but this is an improvement
+  # over the current API for shared plugins
+  # It is best if the codec implements this directly
+  def multi_encode(events)
+    batch = if @has_encode_sync              
+              events.map {|event| [event, self.encode_sync(event)]}
+            else
+              batch = @thread_batches[Thread.current]
+              @multi_encode_mutex.synchronize { batch = @thread_batches[Thread.current] = [] } unless batch
+              batch.clear
+              
+              events.each {|event| self.encode(event) }
+              batch
+            end
+
+    if @on_event != @on_event_proc
```

Comment:
I squint at this and wonder what the meaning of this check is. Can you move this condition `@on_event != @on_event_proc` to a method call with a descriptive name? Something like

```ruby
def multi_encode?
  @on_event != @on_event_proc
end
```